source/mir/base64.d

```diff
@@ -0,0 +1,425 @@
+            sz = 3;
+        }
+
+        decodedByteGroup[0] = (transformed_group >> 24) & 0xff;
+        decodedByteGroup[1] = (transformed_group >> 16) & 0xff;
+        decodedByteGroup[2] = (transformed_group >> 8) & 0xff;
+
+        // Only emit the transformed bytes that we got data for. 
+        appender.put(decodedByteGroup[0 .. sz]);
+    }
+}
+
+/// Test decoding of data which has a length which can be
+/// cleanly decoded.
+unittest
+{
+    {
+        enum data = "QUJD";
+        assert(data.decodeBase64 == "ABC");
+    }
+
+    {
+        enum data = "QQ==";
+        assert(data.decodeBase64 == "A");
+    }
+
+    {
+        enum data = "YSBiIGMgZCBlIGYgZyBoIGkgaiBrIGwgbSBuIG8gcCBxIHIgcyB0IHUgdiB3IHggeSB6";
+        assert(data.decodeBase64 == "a b c d e f g h i j k l m n o p q r s t u v w x y z");
+    }
+
+    {
+        enum data = "LCAuIDsgLyBbICcgXSBcID0gLSAwIDkgOCA3IDYgNSA0IDMgMiAxIGAgfiAhIEAgIyAkICUgXiAmICogKCApIF8gKyB8IDogPCA+ID8=";
+        assert(data.decodeBase64 == ", . ; / [ ' ] \\ = - 0 9 8 7 6 5 4 3 2 1 ` ~ ! @ # $ % ^ & * ( ) _ + | : < > ?");
+    }
+
+    {
+        enum data = "AAA=";
+        assert(data.decodeBase64 == "\x00\x00");
+    }
+
+    {
+        enum data = "AAAABBCC";
+        assert(data.decodeBase64 == "\x00\x00\x00\x04\x10\x82");
+    }
+
+    {
+        enum data = "AA==";
+        assert(data.decodeBase64 == "\x00");
+    }
+    
+    {
+        enum data = "AA/=";
+        assert(data.decodeBase64 == "\x00\x0f");
+    }
+}
+
+/// Test decoding invalid data
+unittest
+{
+    void testFail(const(char)[] input)
+    {
+        bool thrown = false;
+        try {
+            ubyte[] decoded = input.decodeBase64;
+        } catch (Throwable t) {
+            thrown = true;
+        }
+
+        assert(thrown);
+    }
+
+    testFail("===A");
+    testFail("A=");
+    testFail("AA=");
+    testFail("A=AA");
+    testFail("AA=A");
+    testFail("AA=A====");
+    testFail("=AAA");
+    testFail("AAA=QUJD");
+    // This fails because we don't allow extra padding (than what is necessary)
+    testFail("AA======");
+    // This fails because we don't allow padding before the end of the string (otherwise we'd have a side-channel)
+    testFail("QU==QUJD");
+    testFail("QU======QUJD");
+    // Invalid data that's out of the alphabet
+    testFail("!@##@@!@");
+}
+
+/++
+Encode a ubyte array as Base64, returning the encoded value.
++/
+const(char)[] encodeBase64(char PlusChar = '+', char SlashChar = '/')(scope const(ubyte)[] buf) @safe pure
+{
+    import mir.appender : scopedBuffer;
+    // XXX: is a stringBuf more appropriate here?
+    auto app = scopedBuffer!char;
+    encodeBase64!(PlusChar, SlashChar)(buf, app);
+    return app.data.dup;
+}
+
+/++
+Encode a ubyte array as Base64, placing the result onto an Appender.
++/
+void encodeBase64(char PlusChar = '+', char SlashChar = '/', Appender)(scope const(ubyte)[] input,
```

Comment:
No reason to make PlusChar and SlashChar generic params as well as CapitalCase. Make them runtime and camelCase.